File: lib/auth/reset_password_view_model.dart
import 'package:flutter/foundation.dart';
import 'package:diplomovka/auth/reset_password_use_case.dart';
import 'package:firebase_auth/firebase_auth.dart';

class ResetPasswordViewModel extends ChangeNotifier {
  // Use Case je injektovaný v konštruktore (z KROKU 2)
  final ResetPasswordUseCase _resetPasswordUseCase;

  ResetPasswordViewModel(this._resetPasswordUseCase);

  String? _error;
  bool _isLoading = false;

  String? get error => _error;
  bool get isLoading => _isLoading;

  /// Pošle žiadosť o reset hesla a spracuje odpoveď.
  /// Vráti true, ak bolo odoslanie úspešné.
  Future<bool> sendResetEmail(String email) async {
    if (_isLoading) return false;

    _isLoading = true;
    _error = null;
    notifyListeners(); // Upozorníme UI na začiatok načítania

    try {
      // Volanie logiky (Domain Layer)
      await _resetPasswordUseCase.execute(email);
      // Úspech
      return true;
    } on FirebaseAuthException catch (e) {
      // Spracovanie špecifických chýb Firebase
      if (e.code == 'user-not-found') {
        _error = 'Používateľ s týmto e-mailom neexistuje.';
      } else if (e.code == 'invalid-email') {
        _error = 'Neplatný formát e-mailu.';
      } else {
        _error = 'Chyba: ${e.message}';
      }
      return false; // Chyba
    } catch (e) {
      _error = 'Nastala neočakávaná chyba pri odosielaní. Skúste znova.';
      return false; // Chyba
    } finally {
      _isLoading = false;
      notifyListeners(); // Upozorníme UI na koniec načítania
    }
  }

  void clearError() {
    _error = null;
    notifyListeners();
  }
}

File: app/src/main/java/com/example/diplomovka_kotlin/viewmodel/ResetPasswordViewModel.kt
package com.example.diplomovka_kotlin.viewmodel

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import com.example.diplomovka_kotlin.data.services.AuthService
import com.example.diplomovka_kotlin.domain.ResetPasswordUseCase
import com.google.firebase.auth.FirebaseAuthException
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

class ResetPasswordViewModel(application: Application) : AndroidViewModel(application) {

    private val authService = AuthService(application)
    private val resetUseCase = ResetPasswordUseCase(authService)

    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()

    private val _error = MutableStateFlow<String?>(null)
    val error: StateFlow<String?> = _error.asStateFlow()

    suspend fun sendResetEmail(email: String): Boolean {
        if (_isLoading.value) return false
        _isLoading.value = true
        _error.value = null

        return try {
            resetUseCase.execute(email)
            true
        } catch (e: FirebaseAuthException) {
            _error.value = when (e.errorCode) {
                "user-not-found" -> "Používateľ s týmto e-mailom neexistuje."
                "invalid-email"  -> "Neplatný formát e-mailu."
                else -> "Chyba: ${e.message ?: e.errorCode}"
            }
            false
        } catch (e: Exception) {
            _error.value = "Nastala neočakávaná chyba pri odosielaní. Skúste znova."
            false
        } finally {
            _isLoading.value = false
        }
    }

    fun clearError() { _error.value = null }
}
